sdk/eventhub/azure-eventhubs/azure/eventhub/eventprocessor/event_processor.py

```diff
@@ -0,0 +1,189 @@
+                                                            EventPosition(str(offset)))
+            consumers.append(consumer)
+
+            partition_processor = self._partition_processor_factory(
+                eventhub_name=self._eventhub_name,
+                consumer_group_name=self._consumer_group_name,
+                partition_id=partition_id,
+                checkpoint_manager=CheckpointManager(partition_id, self._eventhub_name, self._consumer_group_name,
+                                                     self._id, self._partition_manager)
+            )
+            loop = get_running_loop()
+            task = loop.create_task(
+                _receive(consumer, partition_processor, self._receive_timeout))
+            self._tasks.append(task)
+        try:
+            await asyncio.gather(*self._tasks, return_exceptions=True)
+        finally:
+            await self._partition_manager.close()
+            log.info("EventProcessor %r partition manager is closed", self._id)
+            log.info("EventProcessor %r has stopped", self._id)
+
+
+async def _receive(partition_consumer, partition_processor, receive_timeout):
+    try:
+        while True:
+            try:
+                events = await partition_consumer.receive(timeout=receive_timeout)
+            except asyncio.CancelledError as cancelled_error:
+                log.info(
+                    "PartitionProcessor of EventProcessor instance %r of eventhub %r partition %r consumer group %r "
+                    "is cancelled",
+                    partition_processor._checkpoint_manager._id,
+                    partition_processor._eventhub_name,
+                    partition_processor._partition_id,
+                    partition_processor._consumer_group_name
+                )
+                await partition_processor.process_error(cancelled_error)
+                await partition_processor.close(reason=CloseReason.SHUTDOWN)
+                break
+            except EventHubError as eh_err:
+                reason = CloseReason.LEASE_LOST if eh_err.error == "link:stolen" else CloseReason.EVENTHUB_EXCEPTION
+                log.warning(
+                    "PartitionProcessor of EventProcessor instance %r of eventhub %r partition %r consumer group %r "
+                    "has met an exception receiving events. It's being closed. The exception is %r.",
+                    partition_processor._checkpoint_manager._id,
+                    partition_processor._eventhub_name,
+                    partition_processor._partition_id,
+                    partition_processor._consumer_group_name,
+                    eh_err
+                )
+                await partition_processor.process_error(eh_err)
+                await partition_processor.close(reason=reason)
+                break
+            try:
+                await partition_processor.process_events(events)
+            except asyncio.CancelledError as cancelled_error:
```

Comment:
Isn't this except clause exactly the same as above? Why to separate try/catches for the receive and process calls?

Reply:
Not exactly the same. receive raises EventHubError while process_events doesn't.
Another difference is the handling of error may be different. process_events exceptions are caused by user code. We (eventhus sdk developers) haven't decided whether the handling is different from receive errors. The decision will be after preview 2.
I also think this code can be make shorter. But I would like to do that after we have the decision and finish other higher priority things.

Reply:
Changed in preview 3